examples/dominant-assurance/index.rsh

```diff
@@ -0,0 +1,134 @@
+'reach 0.1';
+'use strict';
+
+/* Based off of https://en.wikipedia.org/wiki/Assurance_contract#Dominant_assurance_contracts
+   Please read that short Wikipedia article to understand the meaning behind participant names
+   and the fields of Product's interact object. */
+
+const InvestmentStructureT = Object({
+  entrepreneurInvestment: UInt, // Entrepreneur's contribution to the product funding
+  entrepreneurProfit: UInt,     // Entrepreneur's profit when the quorum is met
+  investorInvestment: UInt,     // Each investor's contribution to the product funding
+  investorFailProfit: UInt,     // Each investor's profit if the quorum is not met
+  investorQuorum: UInt,         // Number of investors needed to successfully fund the product
+  investmentDuration: UInt,     // How long funding will be open to investors
+  failPayDuration: UInt,        // How long failure pay will be available for investors to claim
+});
+
+export const main = Reach.App(() => {
+  const P = Participant('Product', {
+    investmentStructure: InvestmentStructureT,
+    ready: Fun([], Null),
+  });
+  const E = Participant('Entrepreneur', {});
+  const I = API('Investor', {
+    invest: Fun([], Null),
+    collectFailPay: Fun([], Null),
+  });
+  setOptions({ untrustworthyMaps: true });
+  init();
+
+  const checkInvestmentStructure = (iso) => {
+    const expectedFunds = iso.entrepreneurInvestment
+                        + iso.investorQuorum * iso.investorInvestment;
+    check(iso.entrepreneurProfit <= expectedFunds);
+    check(iso.investorQuorum > 1);
+    check(iso.investorInvestment > 0);
+    check(iso.investorFailProfit > 0);
+  };
+
+  // A participant representing the product being funded specifies how investment will work
+  P.only(() => {
+    const investmentStructure = declassify(interact.investmentStructure);
+    checkInvestmentStructure(investmentStructure);
+  });
+  P.publish(investmentStructure);
+  checkInvestmentStructure(investmentStructure);
+
+  const {
+    entrepreneurInvestment,
+    entrepreneurProfit,
+    investorInvestment,
+    investorQuorum,
+    investorFailProfit,
+    investmentDuration,
+    failPayDuration
+  } = investmentStructure;
+
+  commit();
+
+  // Entrepreneur kicks off the contract by paying their investment plus enough
+  // to compensate investors in the case of failure
+  const starterInvestment = entrepreneurInvestment
+                          + investorFailProfit * (investorQuorum - 1);
+  E.pay(starterInvestment);
```

Comment:
I would put this as part of the original publish on line 45

Reply:
Or at least another API call if you want to separate the entrepreneur from the structure-creator